### ISSUE TITLE
Properties are not reading the correct batched value

### ISSUE THREAD

**Issue by tmat** [Microsoft/msbuild#2249]
Repro:

Create directory structure and empty files like so:
```
A\1.stub
B\2.stub
B\3.stub
```

and run targets file in the same dir as A and B:

```xml
<Project>
  <ItemGroup>
    <_StubFiles Include="$(MSBuildThisFileDirectory)**\*.stub"/>

    <_StubDirs Include="@(_StubFiles->'%(RecursiveDir)')"/>
  </ItemGroup>
  
  <Target Name="Build">
    <PropertyGroup>
      <_ComponentDir>%(_StubDirs.Identity)</_ComponentDir>
      <_ComponentName>$(_ComponentDir.TrimEnd('\'))</_ComponentName>
    </PropertyGroup>
    
    <Message Text=">> %(_StubDirs.Identity) '$(_ComponentDir)' '$(_ComponentName)'"/>
  </Target>
</Project>
```

Actual output:
```
  >> A\ 'B\' 'B'
  >> B\ 'B\' 'B'
```

Expected output:
```
  >> A\ 'A\' 'A'
  >> B\ 'B\' 'B'
```

**Comment by tmat** [comment]
@rainersigwald Is this a feature or a bug? If it is a feature it should be documented here: https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching 
Is there a workaround?

**Comment by tmat** [comment]
FYI @nguerrera

**Comment by nguerrera** [comment]
Oh now that I see the repro, I see the issue as opposed to when we discussed offline.

The property creation and message tasks are being batched separately. So the execution is CreateProperty 2xN times, Message N times. So every time the Message task runs the properties have their final values. You need to declare outputs to trigger target batching instead of task batching.

This produces the expected output:

``` diff
<  <Target Name="Build">
>  <Target Name="Build" Outputs="%(_StubDirs.Identity)">

This distinction between task and target batching is vaguely covered in linked doc.

**Comment by tmat** [comment]
Ah, I see. PropertyGroup is considered a task for the purpose of batching. I did not see that in the doc. It would help to add a sample that sets a property in the target.

**Comment by nguerrera** [comment]
My understanding is that inside a target, PropertyGroup is actually sugar for https://msdn.microsoft.com/en-us/library/63ckb9s9.aspx.

**Comment by tmat** [comment]
Makes sense now. My understanding was that ```%``` finds the enclosing Target or task and I didn't know PG is a task.

**Comment by rainersigwald** [comment]
@nguerrera's analysis is correct, and @tmat is 100% right that the doc could be much, much clearer about this. Repurposing the bug for that doc update.

**Comment by tmat** [comment]
@rainersigwald Thanks.

**Comment by rainersigwald** [comment]
Oh, and to be pedantic it's the individual property elements that are sugar for `CreateProperty`, not the `PropertyGroup`. Which is relevant in this case because batching over the group is what was initially expected.

**Comment by geeeyetee** [comment]
The following behavior looks like a bug to me, but I am noting it here first in case it has already been fixed (I'm afraid I am only testing in Build Engine version 14.0.25420.1) or it is expected behavior (in which case the documentation should explain it in my opinion).

Here are 2 targets that use different approaches to attempt to set the value of the `Max` property to the highest value among the `Number` items:

```
  <Target Name="PopulateMaxViaCondition">
    <ItemGroup>
      <Number Include="2" />
      <Number Include="1" />
    </ItemGroup>
    <PropertyGroup>
        <Max>0</Max>
        <Max Condition="%(Number.Identity) &gt; $(Max)">%(Number.Identity)</Max>
    </PropertyGroup>
    <Message Text="Max: $(Max)"/>
  </Target>
```

```
  <Target Name="PopulateMaxViaPropertyFunction">
    <ItemGroup>
      <Number Include="2" />
      <Number Include="1" />
    </ItemGroup>
    <PropertyGroup>
        <Max>0</Max>
        <Max>$([System.Math]::Max(`%(Number.Identity)`, `$(Max)`))</Max>
    </PropertyGroup>
    <Message Text="Max: $(Max)"/>
  </Target>
```

Output, actual (identical for either target):

> Added Item(s): Number=2
> Added Item(s): Number=1
> Set Property: Max=0
> Set Property: Max=2
> Set Property: Max=1
> Using "Message" task from assembly "Microsoft.Build.Tasks.Core, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
> Task "Message" (TaskId:2)
>   Task Parameter:Text=Max: 1 (TaskId:2)
>   Max: 1 (TaskId:2)
> Done executing task "Message". (TaskId:2)

Output, expected, target `PopulateMaxViaCondition`:

> Added Item(s): Number=2
> Added Item(s): Number=1
> Set Property: Max=0
> Set Property: Max=2
> Using "Message" task from assembly "Microsoft.Build.Tasks.Core, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
> Task "Message" (TaskId:2)
>   Task Parameter:Text=Max: 2 (TaskId:2)
>   Max: 2 (TaskId:2)
> Done executing task "Message". (TaskId:2)

Output, expected, target `PopulateMaxViaPropertyFunction`:

> Added Item(s): Number=2
> Added Item(s): Number=1
> Set Property: Max=0
> Set Property: Max=2
> Set Property: Max=2
> Using "Message" task from assembly "Microsoft.Build.Tasks.Core, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
> Task "Message" (TaskId:2)
>   Task Parameter:Text=Max: 2 (TaskId:2)
>   Max: 2 (TaskId:2)
> Done executing task "Message". (TaskId:2)
